tests/integration/mirrorpeersecret_controller_test.go

```diff
@@ -0,0 +1,615 @@
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+)
+
+// MirrorPeerPlusPlus provides more functionalities to `MirrorPeer`
+type MirrorPeerPlusPlus multiclusterv1alpha1.MirrorPeer
+
+func (mppp MirrorPeerPlusPlus) GetManagedClusters() []clusterv1.ManagedCluster {
+	var managedClusters []clusterv1.ManagedCluster
+	for _, eachItem := range mppp.Spec.Items {
+		managedCluster := clusterv1.ManagedCluster{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: eachItem.ClusterName,
+			},
+			Spec: clusterv1.ManagedClusterSpec{},
+		}
+		managedClusters = append(managedClusters, managedCluster)
+	}
+	return managedClusters
+}
+
+// GetClusterNamespaces returns all the namespaces of each StorageCluster
+// in the MirrorPeer items
+func (mppp MirrorPeerPlusPlus) GetClusterNamespaces() []corev1.Namespace {
+	uniqueNamespaceMap := map[string]struct{}{}
+	for _, eachPeer := range mppp.Spec.Items {
+		uniqueNamespaceMap[eachPeer.ClusterName] = struct{}{}
+	}
+	var namespaceArr []corev1.Namespace
+	for namespace := range uniqueNamespaceMap {
+		namespace := corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: namespace,
+			},
+		}
+		namespaceArr = append(namespaceArr, namespace)
+	}
+	return namespaceArr
+}
+
+// GetDummySourceSecrets generates some dummy source secrets according to the MirrorPeer items
+func (mppp MirrorPeerPlusPlus) GetDummySourceSecrets(prefixName string) []corev1.Secret {
+	var sourceSecrets []corev1.Secret
+	for i, eachPR := range mppp.Spec.Items {
+		secretNN := types.NamespacedName{
+			Name:      fmt.Sprintf("%s-%d", prefixName, i),
+			Namespace: eachPR.ClusterName,
+		}
+		storageClusterNN := types.NamespacedName{
+			Name:      eachPR.StorageClusterRef.Name,
+			Namespace: eachPR.StorageClusterRef.Namespace,
+		}
+		sourceSecret := common.CreateSourceSecret(secretNN, storageClusterNN, []byte("MySecretData1234"))
+		sourceSecrets = append(sourceSecrets, *sourceSecret)
+	}
+	return sourceSecrets
+}
+
+var _ = Describe("MirrorPeerSecret Controller", func() {
+	var (
+		ctx    = context.TODO()
+		mpName = "mirrorpeer-sample"
+		items  = []multiclusterv1alpha1.PeerRef{
+			{
+				ClusterName: "cluster11",
+				StorageClusterRef: multiclusterv1alpha1.StorageClusterRef{
+					Name:      "test-storagecluster-1",
+					Namespace: "test-namespace",
+				},
+			},
+			{
+				ClusterName: "cluster12",
+				StorageClusterRef: multiclusterv1alpha1.StorageClusterRef{
+					Name:      "test-storagecluster-2",
+					Namespace: "test-namespace",
+				},
+			},
+		}
+	)
+
+	var getSourceSecrets = func(ctx context.Context, rc client.Client) ([]corev1.Secret, error) {
+		var sourceList corev1.SecretList
+		var hasSourceLabel client.MatchingLabels = make(client.MatchingLabels)
+		hasSourceLabel[common.SecretLabelTypeKey] = string(common.SourceLabel)
+		err := rc.List(ctx, &sourceList, hasSourceLabel)
+		return sourceList.Items, err
+	}
+
+	When("we start the test", func() {
+		It("should create all the necessary resources", func() {
```

Comment:
This is done in `Before*` block. Let's not use When and It for test preparations.

Reply:
I initially had these changes in the 'BeforeEach` block, but this was giving me errors while running the test.
For example the `AfterEach` block when it deletes all the internally created resources , BeforeEach will give me the error like the resource is being deleted and cannot be created.
I struggled to fix the issue as all the permutations/combinations where blowing in one test or another (always). Writing test took more time than writing the actual functionality... =)

Reply:
`Writing test took more time than writing the actual functionality` - how does that matter?

I am not willing to merge this as it is just because it works and the other approach wasn't working. Why wasn't it working? Why does this work? Maybe this works because it was designed to work instead of test? 

My point is I am not very confident about this. Mayeb we should get more eyes on this.

Reply:
Updated the tests with `BeforeEach` and `AfterEach`. Found the failure part (in `AfterEach` section) and commented it out. Please take a look.